kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/codegen/feel11/CompiledCustomFEELFunction.java

```diff
@@ -33,18 +34,32 @@
 
     private final List<String> parameters;
     private final Function<EvaluationContext, Object> body;
+    private final EvaluationContext ctx;
 
     public CompiledCustomFEELFunction(String name, List<String> parameters, Function<EvaluationContext, Object> body) {
+        this(name, parameters, body, null);
+    }
+
+    public CompiledCustomFEELFunction(String name, List<String> parameters, Function<EvaluationContext, Object> body, EvaluationContext ctx) {
         super( name );
         this.parameters = parameters;
         this.body = body;
+        this.ctx = ctx;
     }
 
     @Override
     public List<List<String>> getParameterNames() {
         return Arrays.asList( parameters );
     }
 
+    public boolean isProperClosure() {
+        return ctx != null;
+    }
+
+    public EvaluationContext getEvaluationContext() {
+        return ctx;
```

Comment:
I have a doubt about this, as I'm struggling to see why it cannot be reused from the `EvaluationContext ctx` in signature of `invoke` just by looking at the code on github. I suggest as tomorrow we are working colocated, we go through this code in person, please?

Reply:
ok. Anyway, the problem is that you can't just invoke a function with a toplevel ctx.
e.g.

```javascript
{
   import: {
      hi: function() "hi "
      hello: function (e) hi() + e
   }
}.import.hello("john")
```

returns null, because `hi()` is unqualified, and current codegen routine is passing `feelExprCtx` which is the toplevel, so `import.hi()` would work, but that's wrong. 

Because `hello(e)` closes over its enclosing context, you can refer to `hi()` unqualified; thus it is a closure, which by definition, holds a reference to its enclosing environment, which is what `EvaluationContext#current()` returns. 

The way `invoke()` is then called is kind of roundabout, but I didn't want to break other stuff. We can talk about this tomorrow though

Reply:
Maybe I think I see what you are saying. in pseudocode-FEEL where I can nest some Java, if I have
```
{
      hi: function() Math.rand(),
      hello: function (e) hi() + e
   }
```
then `hi` being a function every time will be executed, which is what we expect every time we call the `hello` returning every time a different result.

If I had it like:
```
{
      x: Math.rand(),
      hello: function (e) x + e
   }
```
then `x` is (again) analogous to an effectively final value, which is realized the moment `x` is assigned, and this is permanently referenced by `hello`

Also, this on `EvaluationContext` isn't effectively a `clone`?

Reply:
> Also, this on `EvaluationContext` isn't effectively a clone?

Not really, it (should be) a reference to the top of the stack within that context. 

Because a `Deque` was used, we have to copy a slice of that `Deque`. If the stack were implemented in a proper linked list (as in, not `java.util.LinkedList`, but a proper head/tail list), then it would be a single reference to the head of the stack. We can make this change later (Frames are already implemented this way)

Reply:
(as discussed, in summary: would be enough to refactor in `EvaluationContext` to have a pointer to the "peek" instead of a full stack)